### PR TITLE
Implements time delay on rejoining game with ghost roles

### PR DESCRIPTION
**Pull request by coiax**
:cl:
del: Most ghost roles require you to have been dead for five minutes
before it's possible to rejoin the game. Please remember that when you
join as a ghost role, you have no memory of previous lives, even with
the same ghost role.
/:cl:

- Affects drones, swarmers, free golems, anything using the mob_spawn
effect.
- Does not affect any signups for events, or anything that uses a
"popup" window.
- Does not affect CTF.
- Does effect observers that have never died, if you lack a body, your
"timeofdeath" is when your ghost was created.

I am a big advocate for ghost roles, and I'm well aware that many people
just like observing at round start and playing drone or free golem. This
PR is not aimed at them.

Ghost roles are NOT a "free respawn" for you to pop into a new body and
immediately continue whatever you were doing before you died. You should
not (and now cannot) respawn as a drone constantly every time you are
killed, you should not become a wave of golems attacking the cult base
that all suddenly know exactly where the cult base is and feel very
strongly about killing cultists.

Behaviour like this gives people who actually enjoy ghost roles on their
own merits a bad name.

This change will allow some subsequent changes that I'd like to
implement that would probably be disallowed if instant ghost role
respawning was still present.

### CODE REVIEW

**Comment by Tlaltecuhtli** [comment]
>write a paper with indications to where the cult base is
problem solved

**Comment by Tlaltecuhtli** [comment]
does it affect roles like marauder and cult ghost?

**Comment by Repukan** [comment]
This nerfs swarmers since their main strenght is being able to quickly come back.

**Comment by FantasticFwoosh** [comment]
Ok we get it, we know ashwalkers dunked you again just stab us from the front and start reverting spammy ghost roles you dont like already instead of this shady shit.

**Comment by FantasticFwoosh** [comment]
Besides what cyberboss might think of my deleted comment, this also nerfs ashwalkers too who are also antagonists that rely on quick co-operation. Seems like unnecessary sweeping nerfs when you can just apply the respawn timers individually.

**Comment by MoreRobustThanYou** [comment]
Port paradise's respawn system. Basically ghosts have a "Respawnability: yes/no" var, which affects if they can take ghost roles / be cloned / borged / etc. If they suicide/ghost early in the round, it gets set to no. However, ghosts can also activate antag hud in exchange for giving up their respawnability

File: code/modules/mob/dead/observer/observer.dm [hunk 1]
```diff
@@ -826,3 +831,20 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 		spawners_menu = new(src)
 
 	spawners_menu.ui_interact(src)
+
+/mob/dead/observer/proc/respawn_check(var/thing = "cancer rat", var/respawn_time = 2 MINUTES, feedback = TRUE)
```
Comment (Cyberboss) [inline code/modules/mob/dead/observer/observer.dm]: remove var/s

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -826,3 +831,20 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
+	if(!client)
+		return FALSE
```
Comment (Cyberboss) [inline code/modules/mob/dead/observer/observer.dm]: I don't see a path where this check can proc?

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Potato-Masher** [comment]
My main problem with this is it doesn't really fix the issue of people treating it as a fresh respawn. It's just putting a 5 min timer before they can get their potential respawn. Honestly, I would think stricter admin reinforcement of the rules regarding this would be a better fix, and it wouldn't screw over antags that require ghosts to take control. If there are no current ghosts and I disintegrate someone as a Servant of Ratvar, it's gonna take an extra 5 mins to get that maurader shell I want to let loose on the station to turn on, which is a TON of wasted time that the shell just sits there.

In short, I think we just need to enforce the rules a little better rather than putting a time restriction on things.

**Comment by arsserpentarium** [comment]
You can fix swatmers fairly easy.
By respawn_time var. So ashwalkers and swarmers will have low time.

**Comment by MoreRobustThanYou** [comment]
This is the best picture I could get with my crappy LG phone

img_20171230_204825_1

**Comment by BeeSting12** [comment]
Primary round antagonist ghost roles such as cultist ghost roles should not have a timer on them. Swarmers too. I don't see an issue with all the ghost roles being slapped with this though, getting attacked by the same guy in three different ash walker bodies is kinda dumb and it will force them to be more careful with their one life for the next five minutes.

**Comment by KorPhaeron** [comment]
This kinda screws over ash walkers who are encouraged to swarm and die

**Comment by KorPhaeron** [comment]
I mean I get the problem you are trying to solve but this mostly seems like it will lead to more frustration for everyone while attempting to deal with a rules issue

**Comment by coiax** [comment]
I think people abuse ghost roles to get back into the round, it's not noticed or reported enough. And the use of metaknowledge isn't allowed when people die and get cloned, so why is it permissible with ghost roles?

People behaving like this gives people who ACTUALLY play ghost roles on their own merit a bad name.

**Comment by coiax** [comment]
Ash walkers aren't "encouraged" to do anything. They have a whole system of botany and milking the damn slupers up for healing.

**Comment by KorPhaeron** [comment]
Those suck and they refund parts of their lives via suicide

They are individually outmatched by just about anything as well

I also think itd be duller as a miner/security if you shot two ash walkers and then a nest with 6 eggs just sat there defenseless

**Comment by coiax** [comment]
Then we can set a lower respawn timer for them then. Nothing's set in stone.

Maybe two minutes for ash walkers and swarmers?

**Comment by coiax** [comment]
I want to make a variety of changes, and I don't want the denial reason because "lol ghost roles get gud don't die"

**Comment by KorPhaeron** [comment]
What are your planned changes after this?

**Comment by FantasticFwoosh** [comment]
Like i alluded to _bluntly_ in my deleted comment it does appear you have grievance @coiax with ashwalkers particularly without the desire to address it directly.

Same goes for you with "_**git gud don't die**_" if you actually find them threatening like @KorPhaeron says where they are easily defeated, but i take the acknowledgement given my prior experience with them that they can snowball into a problem (and that the current ruin is unbalanced for starting supplies)

This change if implemented hurts too many roles & future ghost roles as is without more development of what kind of mechanical gameplay change you want to implement when the affected roles all function differently, free-golems are slow to churn out new shells at first but no scientist in xenobio or RnD wants to wait 5 minutes for a posibrain or slave golem core to be ready.

**Comment by coiax** [comment]
This PR is not about ash walkers. We can disable any form of minimum death time for them if that's what's needed.

I'd like to talk to Kor personally about this though, without every comment going through a sieve.

**Comment by epochayur** [comment]
>You should not (and now cannot) respawn as a drone constantly every time you are killed

remind me again why this is supposedly a bad thing?

**Comment by Iamgoofball** [comment]
@KorPhaeron @coiax can you guys implement a bypass option for this, dwarves would need a free pass on this

**Comment by Bawhoppen** [comment]
No they wouldn't. Not any more than any other ghost role.

**Comment by ExcessiveUseOfCobblestone** [comment]
>No they wouldn't. Not any more than any other ghost role.

Ashwalkers need it anyways.

**Comment by Cyberboss** [comment]
Ping for reopen when we're sure what we're doing about this
